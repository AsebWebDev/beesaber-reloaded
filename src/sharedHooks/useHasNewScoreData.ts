import { skipToken } from '@reduxjs/toolkit/dist/query';
import { useEffect, useState } from 'react';

import { useGetFullPlayerQuery } from '@/api/services/apiPlayer/apiPlayer';
import { useGetUserDataQuery } from '@/api/services/apiUser/apiUser';

const useHasNewScoreData = (userId: string): boolean => {
  const [scoreSaberId, setScoreSaberId] = useState<string | undefined>(
    undefined
  );
  const [hasNewScores, setHasNewScores] = useState<boolean>(false);

  const { data } = useGetUserDataQuery(userId);

  const _scoreSaberId = data?.playerInfo?.playerId;
  const totalPlayCount = data?.totalPlayCount;

  useEffect(() => {
    if (_scoreSaberId !== undefined) setScoreSaberId(_scoreSaberId);
  }, [_scoreSaberId]);

  const { data: scoreSaberData } = useGetFullPlayerQuery(
    scoreSaberId ?? skipToken
  );

  useEffect(() => {
    const totalPlayCountSS = scoreSaberData?.scoreStats.totalPlayCount;

    if (
      totalPlayCount !== undefined &&
      totalPlayCountSS !== undefined &&
      totalPlayCount < totalPlayCountSS
    )
      setHasNewScores(true);
  }, [scoreSaberData]);

  return hasNewScores;
};

export default useHasNewScoreData;
