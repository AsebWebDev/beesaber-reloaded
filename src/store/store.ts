import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/dist/query/react';

import { apiInitialState } from '@/api/api';
import { apiAuth } from '@/api/services/apiAuth/apiAuth';
import { apiPlayer } from '@/api/services/apiPlayer/apiPlayer';
import { apiUser } from '@/api/services/apiUser/apiUser';

import { initialState as appStatusInitialState } from './reducer/appStatusReducer';
import { initialState as notificationsInitialState } from './reducer/notificationsReducer';
import rootReducer from './reducer/rootReducer';
import { initialState as userDataInitialState } from './reducer/userDataReducer';

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(
      apiAuth.middleware,
      apiPlayer.middleware,
      apiUser.middleware
    ),
  devTools: true,
});

setupListeners(store.dispatch);

const initialState = {
  appStatus: { ...appStatusInitialState },
  notifications: { ...notificationsInitialState },
  userData: { ...userDataInitialState },
  api: { ...apiInitialState },
};

export { initialState };

export default store;

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
