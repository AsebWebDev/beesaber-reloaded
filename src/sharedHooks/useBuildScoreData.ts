/* eslint-disable no-console */
import { useEffect, useState } from 'react';
import { toast } from 'react-toastify';

import getAllScores from '@/api/services/helper/getAllScores';
import { useAppDispatch } from '@/store/hooks';
import { userIsFetchingData } from '@/store/reducer/appStatusReducer';

import type { ScoreData } from '@/../sharedTypes';

/**
 * Fetch and parse all score data of a user
 * @param {String} id - ScoreSaber id
 * @param {number} threshold - optional threshold for number of score pages (1 page = 8 scores)
 * @returns {ScoreData} score data object with recent scores, top scores and a list of
 * score hashes
 */
const useBuildScoreData = (
  id?: string,
  threshold?: number
): ScoreData | undefined => {
  if (id === undefined) return undefined;
  const dispatch = useAppDispatch();
  const [isFetching, setIsFetching] = useState(false);
  const [scoreData, setScoreData] = useState<ScoreData>({
    scoredSongsHashes: [],
    scoresRecent: [],
    scoresTop: [],
  });

  useEffect(() => {
    setIsFetching(true);
    const fetchData = async () => {
      try {
        const resp = await toast.promise<ScoreData>(
          getAllScores(id, threshold),
          {
            pending: 'Updating scores ...',
            success: 'We updated all Scores ðŸ‘Œ',
            error: 'There has been an issue your scores ðŸ¤¯',
          }
        );

        setScoreData(resp);
        setIsFetching(false);
      } catch (error: unknown) {
        console.log(
          'ðŸš€ ~ file: useBuildScoreData.ts ~ line 34 ~ fetchData ~ error',
          error
        );
        setIsFetching(false);
      }
    };

    void fetchData();
  }, [id]);

  dispatch(userIsFetchingData({ status: isFetching }));

  return scoreData;
};

export default useBuildScoreData;
