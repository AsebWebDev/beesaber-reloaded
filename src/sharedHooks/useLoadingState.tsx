import { skipToken } from '@reduxjs/toolkit/dist/query/react';

import { useGetUserDataQuery } from '@/api/services/apiUser/apiUser';
import { useAppSelector } from '@/store/hooks';
import { selectUserId } from '@/store/reducer/userDataReducer';

const useLoadingState = (): boolean => {
  const userId = useAppSelector(selectUserId);
  const { isFetching: isFetchingUserData } = useGetUserDataQuery(
    userId ?? skipToken
  );

  return isFetchingUserData;
};

export default useLoadingState;
