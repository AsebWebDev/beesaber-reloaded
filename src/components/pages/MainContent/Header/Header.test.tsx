import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import * as hooks from '@/sharedHooks/useLoadingState';
import { initialState as initialStore } from '@/store/store';

import Header from './Header';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Header', () => {
  it.each([true, false])(
    'should match snapshot, when is component is fetching=%s',
    (isFetching) => {
      const extendedStore = Object.assign(initialStore, {
        userData: {
          _id: '2141234124',
        },
      });

      jest.spyOn(hooks, 'default').mockReturnValue(isFetching);

      const { container } = render(
        <Provider store={mockStore(extendedStore)}>
          <Header />
        </Provider>
      );

      expect(container.firstChild).toMatchSnapshot();
    }
  );
});
