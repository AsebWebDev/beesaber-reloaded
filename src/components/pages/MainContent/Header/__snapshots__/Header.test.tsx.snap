// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot, when is component is fetching=false 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 2rem;
}

<div
  class="c0"
/>
`;

exports[`Header should match snapshot, when is component is fetching=true 1`] = `
.c1 .ldio-do9usc7wgsf div {
  position: absolute;
  border-width: 0.3125rem;
  border-style: solid;
  opacity: 1;
  border-radius: 50%;
  -webkit-animation: ldio-do9usc7wgsf 1s cubic-bezier(0,0.2,0.8,1) infinite;
  animation: ldio-do9usc7wgsf 1s cubic-bezier(0,0.2,0.8,1) infinite;
}

.c1 .ldio-do9usc7wgsf div:nth-child(1) {
  border-color: #ffeb00;
  -webkit-animation-delay: 0s;
  animation-delay: 0s;
}

.c1 .ldio-do9usc7wgsf div:nth-child(2) {
  border-color: #0000FF;
  -webkit-animation-delay: -0.5s;
  animation-delay: -0.5s;
}

.c1 .loadingio-spinner-ripple-rt8n616prbh {
  width: 3.8125rem;
  height: 3.8125rem;
  display: inline-block;
  overflow: hidden;
  background: transparent;
}

.c1 .ldio-do9usc7wgsf {
  width: 100%;
  height: 100%;
  position: relative;
  -webkit-transform: translateZ(0) scale(0.61);
  -ms-transform: translateZ(0) scale(0.61);
  transform: translateZ(0) scale(0.61);
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
}

.c1 .ldio-do9usc7wgsf div {
  box-sizing: content-box;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 2rem;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="loadingio-spinner-ripple-rt8n616prbh"
    >
      <div
        class="ldio-do9usc7wgsf"
      >
        <div />
        <div />
      </div>
    </div>
  </div>
</div>
`;
